fem/intrules_cut.cpp

```diff
@@ -0,0 +1,1585 @@
+          || LvlSet->Eval(Trafo, ipD) > -1e-12)
+      {
+         element_int = true;
+      }
+
+      Array<int> faces;
+      Array<int> cor;
+      mesh->GetElementFaces(elem, faces, cor);
+      FaceElementTransformations* FTrans =
+         Trafo.mesh->GetFaceElementTransformations(faces[face]);
+      FTrans->SetIntPoint(&(FaceIP[0]));
+
+      Vector normal(Trafo.GetDimension());
+      normal = 0.;
+      if (face == 0 || face == 5)
+      {
+         normal(2) = 1.;
+      }
+      if (face == 1 || face == 3)
+      {
+         normal(1) = 1.;
+      }
+      if (face == 2 || face == 4)
+      {
+         normal(0) = 1.;
+      }
+      if (face == 0 || face == 1 || face == 4)
+      {
+         normal *= -1.;
+      }
+
+      for (int ip = 0; ip < FaceIP.Size(); ip++)
+      {
+         DenseMatrix shape;
+         Vector point(3);
+         IntegrationPoint ipoint;
+         FTrans->Transform(FaceIP[ip], point);
+         Trafo.TransformBack(point, ipoint);
+         BasisAD3D(ipoint, shape);
+
+         for (int dof = 0; dof < nBasis; dof++)
+         {
+            Vector adiv(Trafo.GetSpaceDim());
+            shape.GetRow(dof, adiv);
+            RHS(dof) += (adiv * normal) * FaceWeights(ip, faces[face]);
+         }
+      }
+   }
+
+   // If the element is intersected, integrate over the cut surface (using the
+   // already computed rule) and solve the matrix for the weights.
+   if (element_int && !interior)
+   {
+      H1_FECollection fec(lsOrder, 3);
+      FiniteElementSpace fes(Tr.mesh, &fec);
+      GridFunction LevelSet(&fes);
+      LevelSet.ProjectCoefficient(*LvlSet);
+      mesh->GetElementTransformation(elem, &Trafo);
+
+      const FiniteElement* fe = fes.GetFE(elem);
+      Vector normal(Trafo.GetDimension());
+      Vector gradi(Trafo.GetDimension());
+      DenseMatrix dshape(fe->GetDof(), Trafo.GetDimension());
+      Array<int> dofs;
+      fes.GetElementDofs(elem, dofs);
+
+      // Integrate over the cut surface using the already computed rule.
+      for (int ip = 0; ip < sir->GetNPoints(); ip++)
+      {
+         Trafo.SetIntPoint(&(sir->IntPoint(ip)));
+
+         normal = 0.;
+         fe->CalcDShape(sir->IntPoint(ip), dshape);
+         for (int dof = 0; dof < fe->GetDof(); dof++)
+         {
+            dshape.GetRow(dof, gradi);
+            gradi *= LevelSet(dofs[dof]);
+            normal += gradi;
+         }
+         normal *= (-1. / normal.Norml2());
+
+         DenseMatrix shapes;
+         BasisAD3D(sir->IntPoint(ip), shapes);
+
+         for (int dof = 0; dof < nBasisVolume; dof++)
+         {
+            Vector adiv(Trafo.GetSpaceDim());
+            shapes.GetRow(dof, adiv);
+            RHS(dof) +=  (adiv * normal) * sir->IntPoint(ip).weight;
+         }
+      }
+
+      // solve the underdetermined linear system
+      Vector temp(nBasisVolume);
+      Vector temp2(ir.GetNPoints());
+      VolumeSVD->LeftSingularvectors().MultTranspose(RHS, temp);
```

Comment:
@J-Pi I'm hitting an assert here in `debug` mode, as the `VolumeSVD` has incompatible size with `RHS`. Seems there's a bug since `RHS` is of size `nBasis` instead of `nBasisVolume`.

Also the loop on 1131 goes over volumetric basis `(nBasisVolume)`, but the one on 1087 goes over the surface basis `(nBasis)`.

Could you take a look and confirm?

Reply:
I will take a closer look tomorrow, but it seems that you are right.

Reply:
You are completly right. That is a bug. It mus be `nBasisVolume`